### PR TITLE
Using multiple stations and multiple levels for aggregate option when infilling missing values

### PR DESCRIPTION
**Pull request by shadrackkibet**
@rdstern I am putting up this for testing. I think I have been able to implement these methods for multiple stations and also the Average option can now use a multiple receiver, which means you can use multiple levels at a time. Please also check the start and end options under constant option if it works as discussed. I also added the `sample` as a function to be selected under average option dropdown (I think we need to rename this option to accommodate sample function but I am not sure what to call it now).

**NOTE** This is not ready for code checking.

### CODE REVIEW

**Comment by shadrackkibet** [comment]
@rdstern I also respond quickly. Perhaps this will shed more light in your further testing.

1) I assumed when you want to use multiple stations, for now, you include it in the multiple receiver. I will try to see how to incorporate station receiver for this option.
2) This control is still there but below the selector. See below.
image
3) With my quick testing, I got an impression it seems to work using multiple factors directly. Anyway, I will still check the results keenly. I filtered to remain with missing values and then used multiple factors etc and it gave me different results. 
4)  Which error did you get? What exactly did you do to get an error?

**Comment by shadrackkibet** [comment]
@rdstern this works well for me. I tried with one factor then two and gave me the same values all through for the three functions (mean, median and sample). Perhaps my testing is not good enough. I would like to see how you test it. I think with a skype call things will be clear, lets do that at some point tomorrow.

**Comment by shadrackkibet** [comment]
@rdstern sampling should now work discussed.

**Comment by rdstern** [comment]
@shadrackkibet I could approve now if pushed, but there are some small things to (at least) try perhaps.  In bold below.
On the different options:
a) Interpolate seems to work ok
b) In Average **_Could you change the name of Average to Typical._**  Then the Sample option remains appropriate here.
c) Constant works ok.  Here there is the option of omitting the missing at the top and bottom.
d) Spline works ok.  It shows the importance (in the longer run) of being able to omit missing values at the top and bottom.  That's not a job for now.
image
e) **_Copy doesn't seem to work - at least not as I expected it to!_**  See below with the default setting of copy from above.

image

(The first missing value seems to be the one that is 2 from below - and from the bottom to infill the top 2 rows.  The second is the value 1 from below.  So the 2 values inserted are not the same?)

f) The structts doesn't work.  I tried making the yield1 into a ts object (without the missing values at the top.  Then it should work and would be quite an attractive option.  There is a problem with the Mozambique data that zoo objects (perhaps ts as well) don't like multiple stations.  In the future we could do your zoo (or zoo.reg as in your existing code, for a single station with the Date field, because zoo(tmin,Date) is ok.  Then I am very curious to see the results.  So, for now I would like to try this option where I make the variable into a ts or zoo object outside the dialogue.  But in the code you currently have hard wired frequency = 2, which doesn't work at all. **_Could you make this code simpler_** to try with variables that have been pre-defined for now.

**Comment by shadrackkibet** [comment]
> Copy doesn't seem to work - at least not as I expected it to! See below with the default setting of copy from above.

@rdstern I suggest changing this label to **Copy from below** by default it is unchecked which means it copies from above.

**Comment by rdstern** [comment]
@shadrackkibet agreed.

**Comment by shadrackkibet** [comment]
@rdstern I have tried to look closely into `na.structTS` from `zoo` and as far as I understand this might not work for us. It doesn't seem to cope with "our" daily data because `frequency` is always 1? There is an interesting discussion from StackOverflow about it where someone had a similar problem see [https://stackoverflow.com/questions/55573930/error-in-imputing-daily-time-series-with-zoo-package](https://stackoverflow.com/questions/55573930/error-in-imputing-daily-time-series-with-zoo-package). The solution provided there is to use `imputeTS` package for time series data and seems to cope well with data of `frequency=1`, looking at it from a distance it has more methods to impute missing values which might be interesting for us to try. More importantly, it also includes KalmanSmoothing `na.kalman(x, model = "StructTS", smooth = TRUE) ` which works like `na.structTS` from `zoo`? but not exactly with the same results! I wonder if you looked into this package (`imputeTS`) before and you might have something to report. If I remember well you might have mentioned it sometime. 

My initial impression is that we will use `imputeTS` here. This might also be a good task for one of the interns to examine the package generally and see how the other methods work i.e `na.interpolation(), na.locf(), na.ma(), na.kalman(), na.seadec(), na.seasplit()`. There is a paper published about the package [here](https://journal.r-project.org/archive/2017/RJ-2017-009/index.html).

**Comment by shadrackkibet** [comment]
> a) Interpolate seems to work ok

Nothing TODO.

> b) In Average Could you change the name of Average to Typical. Then the Sample option remains appropriate here.

Now sorted.

> c) Constant works ok. Here there is the option of omitting the missing at the top and bottom.

Nothing TODO?

> d) Spline works ok. It shows the importance (in the longer run) of being able to omit missing values at the top and bottom. That's not a job for now.

Nothing to do for now.

> e) Copy doesn't seem to work - at least not as I expected it to! See below with the default setting of copy from above.

Should be working as expected now!

> f) The structts doesn't work. I tried making the yield1 into a ts object (without the missing values at the top. Then it should work and would be quite an attractive option. There is a problem with the Mozambique data that zoo objects (perhaps ts as well) don't like multiple stations. In the future we could do your zoo (or zoo.reg as in your existing code, for a single station with the Date field, because zoo(tmin,Date) is ok. Then I am very curious to see the results. So, for now I would like to try this option where I make the variable into a ts or zoo object outside the dialogue. But in the code you currently have hard wired frequency = 2, which doesn't work at all. Could you make this code simpler to try with variables that have been pre-defined for now.

Discussion in progress.

File: instat/static/InstatObject/R/stand_alone_functions.R [hunk 1]
```diff
@@ -1275,4 +1275,10 @@ in_top_n <- function(x, n = 10, wt, fun = sum) {
...
   else dat <- dat %>% count(x, sort = TRUE, name = "fq")
   return(x %in% dat$x[1:n])
-}
\ No newline at end of file
+}
+
+summary_sample <- function(x, size, replace = FALSE){
+  if(length(x)==0){return(NA)}
+  else if(length(x)==1){return(x)}
+  else{sample(x = x, size = size, replace = replace)}
```
[the remark this answers is not in the document]
Comment (shadrackkibet) [reply instat/static/InstatObject/R/stand_alone_functions.R:1283]: R returns the last evaluated value without using `return` function. R community claims it is faster than using `return()` explicitly. More detailed explanation [here](https://stackoverflow.com/questions/11738823/explicitly-calling-return-in-a-function-or-not)

**Pull request merged**